### PR TITLE
Add initial extractor implementation

### PR DESCRIPTION
**Pull request by glesica**
It turned out to be easier to use the excellent IRSx documentation in combination with XML XPath selectors than to use the IRSx library itself, so that's what I did.

Right now the app just grabs a handful of filings from an index in the `fixtures/` directory. There is still work to be done (see the TODOs littered about). I'll ticketize those when I get a chance.

Once someone else gets it running, let's merge it.

### CODE REVIEW

File: extractor/extractor/cache.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+
+class DirectoryCache(Cache):
+    """
+    A simple on-disk cache that saves the filing XML files to the given
+    directory.
+
+    TODO: Accept a path template to use with object IDs
+    TODO: Opened files need to be closed
+    TODO: OMG so much edge case and error checking, seriously
```
Comment (smai-f) [inline extractor/extractor/cache.py:45]: lol

File: extractor/extractor/filing.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+    def __init__(
+        self, xml_file: TextIO,
+    ):
+        self._namespaces = {
+            "efile": "http://www.irs.gov/efile",
+        }
+
+        content = xml_file.read()
+        while not content.startswith("<"):
```
Comment (smai-f) [inline extractor/extractor/filing.py:44]: what about "<" makes it the base case here?

File: extractor/extractor/filing.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+            content = content[1:]
+        content = content.encode("utf-8")
+        _logger.info(f"XML: '{content[:40]}'")
```
Comment (smai-f) [inline extractor/extractor/filing.py:47]: what's at the 40th index?

File: extractor/extractor/filing.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    @property
+    def employee_count(self) -> Optional[int]:
+        return self.get_int("/IRS990/TotalEmployeeCnt")
+
+    @property
+    def principal_officer_name(self) -> Optional[str]:
+        return self.get_str("/IRS990/PrincipalOfficerNm")
+    
+    @property
+    def us_address(self) -> Optional[str]:
```
Comment (smai-f) [inline extractor/extractor/filing.py:72]: possible TODO - would it be up in here that we'd add the conversion of the address to long/lat to a field if it's not already populated?

File: extractor/fetch_index.sh [hunk 5]
```diff
@@ -0,0 +1,28 @@
+#!/usr/bin/env sh
```
Comment (smai-f) [inline extractor/fetch_index.sh:1]: Just curious, have you successfully run this on a more general query with hella data?

**Comment by smai-f** [comment]
QA +1 - I ran the extractor locally as-is and also added additional filing fields after running it once. The output is added to .extractor. The cache also works.

File: extractor/fetch_index.sh [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+set -e
+
+STATE=mt
+# TODO: Allow multiple years
+YEAR=2019
+
+# TODO: Convert to a makefile to avoid redundant downloads
+# TODO: Check for irsx_index in the PATH
```
Comment (smai-f) [inline extractor/fetch_index.sh:10]: For posterity: https://github.com/jsfenfen/990-xml-reader/blob/master/README.md#irsx_index

**Comment by smai-f** [comment]
Just checking for understanding, please correct me if I'm wrong: 

The current `index.csv` file in the fixtures is just stuff you manually grabbed and threw in as a proof of concept. In the future, this will be a static asset that we download years' worth of and put somewhere. Would we do something like download years 2005 (or whatever) - 2018 and cache everything, but for years 2018 - 2020, re-download these years every time and re-run the extractor to grab updates and additions in case new stuff is available? Or does it not change that often? 

The `Index` is our interface between csv to getting the full xml data for a particular 990 filing.
The `Query` is our interface between the `Filing` and xml.
The `Filing` is what the modules will actually use to tell the extractor what it cares about from the data. All will be magical and auto-completed by the time anyone writing a module has to interface with the extractor.

How will the modules talk to the extractor to tell it what they want?

File: extractor/extractor/filing.py (hunk 2 above)
Comment (glesica) [reply extractor/extractor/filing.py:44]: For whatever reason, the XML file comes back with a unicode sequence at the beginning that is apparently supposed to identify the content as UTF-8 encoded, but `lxml`, the Python XML library I'm using, apparently hates that. So I just strip off everything at the beginning until the first tag, which starts with "<". It's weird and I don't totally understand it. I bet Sean would.

File: extractor/extractor/filing.py (hunk 3 above)
Comment (glesica) [reply extractor/extractor/filing.py:47]: Nothing in particular, just trying to get a little bit of the text but without barfing too much into the terminal.

File: extractor/extractor/filing.py (hunk 4 above)
Comment (glesica) [reply extractor/extractor/filing.py:72]: Yeah, good call, the lat/lon stuff will need to plug in somewhere around here.

File: extractor/fetch_index.sh (hunk 5 above)
Comment (glesica) [reply extractor/fetch_index.sh:1]: Negative. May need to parallelize some stuff with `multiprocessing` or `threading` or some other Python library.

**Comment by glesica** [comment]
> The current index.csv file in the fixtures is just stuff you manually grabbed and threw in as a proof of concept.

Accurate, we'll want some mechanism for grabbing indices.

> Would we do something like download years 2005 (or whatever) - 2018 and cache everything, but for years 2018 - 2020, re-download these years every time and re-run the extractor to grab updates and additions in case new stuff is available?

Something like that, we could build in a way to specify a "timeout" for the cache for each index maybe. Then we could tweak that once we see how it works.

> How will the modules talk to the extractor to tell it what they want?

I was thinking maybe a sort of templated JSON file:

```
{
   "address": "<us_address>",
   ...
}
```

So in this case the extractor would grab each one and use it as a template to barf out one or more JSON files with the `<...>` bits replaced with the data?

**Pull request merged**